src/chainlink/oracle/ChainlinkOracleImpl.sol

```diff
@@ -0,0 +1,271 @@
+    /// functions - private & internal
+    /// -----------------------------------------------------------------------
+
+    function _setPairDetails(SetPairDetailParams[] calldata params_) internal {
+        $_pairDetails._set(_convert, params_);
+    }
+
+    /// -----------------------------------------------------------------------
+    /// functions - private & internal - views
+    /// -----------------------------------------------------------------------
+
+    function _getQuoteAmount(QuoteParams calldata quoteParams_) internal view returns (uint256) {
+        ConvertedQuotePair memory cqp = quoteParams_.quotePair._convert(_convert);
+
+        if (cqp.cBase == cqp.cQuote) {
+            return quoteParams_.baseAmount;
+        }
+
+        PairDetail memory pd = $_pairDetails._get(cqp);
+        if (pd.path.length == 0) {
+            revert InvalidPair_FeedNotSet(quoteParams_.quotePair);
+        }
+
+        Feed[] memory feeds = pd.path.getFeeds();
+        uint256 feedLength = feeds.length;
+
+        uint256 price = WAD;
+        for (uint256 i; i < feedLength;) {
+            uint256 answer = _getFeedAnswer(feeds[i]);
+            if (feeds[i].mul) {
+                price = price.mulWadDown(answer);
+            } else {
+                price = price.divWadDown(answer);
+            }
+            unchecked {
+                ++i;
+            }
+        }
+        if (pd.inverted) price = WAD.divWadDown(price);
+        return _convertPriceToQuoteAmount(price, quoteParams_);
+    }
+
+    function _getFeedAnswer(Feed memory feed_) internal view returns (uint256) {
+        (
+            , /* uint80 roundId, */
+            int256 answer,
+            , /* uint256 startedAt, */
+            uint256 updatedAt,
+            /* uint80 answeredInRound */
+        ) = feed_.feed.latestRoundData();
+
+        if (updatedAt < block.timestamp - feed_.staleAfter) {
+            revert StalePrice(feed_.feed, updatedAt);
+        }
+        if (answer < 0) {
+            revert NegativePrice(feed_.feed, answer);
+        }
+
+        if (feed_.decimals <= 18) {
+            return uint256(answer) * 10 ** (18 - feed_.decimals);
+        } else {
+            return uint256(answer) / 10 ** (feed_.decimals - 18);
+        }
+    }
+
+    function _convertPriceToQuoteAmount(uint256 price_, QuoteParams calldata quoteParams_)
+        internal
+        view
+        returns (uint256 finalAmount)
+    {
+        uint8 baseDecimals = quoteParams_.quotePair.base._decimals();
+        uint8 quoteDecimals = quoteParams_.quotePair.quote._decimals();
+
+        if (18 > quoteDecimals) {
+            finalAmount = price_ / (10 ** (18 - quoteDecimals));
+        } else if (18 < quoteDecimals) {
+            finalAmount = price_ * (10 ** (quoteDecimals - 18));
+        } else {
+            finalAmount = price_;
+        }
+        return finalAmount * quoteParams_.baseAmount / 10 ** baseDecimals;
```

Comment:
i think i might re-arrange this a bit:
```
        finalAmount = price_ * quoteParams_.baseAmount / 10 ** baseDecimals;
        if (18 > quoteDecimals) {
            finalAmount = price_ / (10 ** (18 - quoteDecimals));
        } else if (18 < quoteDecimals) {
            finalAmount = price_ * (10 ** (quoteDecimals - 18));
        }
```
this is fewer lines + it keeps the calc at 18 decimals until the very end; otherwise we move (likely?) down to `quoteDecimals` before the final change. wdyt?